src/sentry/rules/history/backends/postgres.py

```diff
@@ -42,15 +46,21 @@ def fetch_rule_groups_paginated(
         cursor: Cursor | None = None,
         per_page: int = 25,
     ) -> CursorResult[Group]:
+        filtered_history = RuleFireHistory.objects.filter(
+            rule=rule,
+            date_added__gte=start,
+            date_added__lt=end,
+        )
+
+        # subquery that retrieves row with the largest date in a group
+        group_max_dates = filtered_history.filter(group=OuterRef("group")).order_by("-date_added")[
+            :1
+        ]
         qs = (
-            RuleFireHistory.objects.filter(
-                rule=rule,
-                date_added__gte=start,
-                date_added__lt=end,
-            )
-            .select_related("group")
-            .values("group")
-            .annotate(count=Count("id"), last_triggered=Max("date_added"))
+            filtered_history.values("group")
+            .annotate(count=Count("group"))
+            .annotate(event_id=Subquery(group_max_dates.values("event_id")))
+            .annotate(last_triggered=Subquery(group_max_dates.values("date_added")))
         )
```

Comment:
I'm a bit worried that this query is going to be expensive. Could you print out what it looks like and post it here? (you can do `print(qs.query)` to get the sql)

Reply:
Looks like this after replacing the second subquery with max
```
SELECT   "sentry_rulefirehistory"."group_id",
         Count("sentry_rulefirehistory"."group_id") AS "count",
         (
                  SELECT   u0."event_id"
                  FROM     "sentry_rulefirehistory" U0
                  WHERE    (
                                    u0."date_added" >= 2022-09-21 00:59:26.899530+00:00
                           AND      u0."date_added" < 2022-09-28 00:59:26.899530+00:00
                           AND      u0."rule_id" = 2
                           AND      u0."group_id" = ("sentry_rulefirehistory"."group_id"))
                  ORDER BY u0."date_added" DESC limit 1) AS "event_id",
         max("sentry_rulefirehistory"."date_added")      AS "last_triggered"
FROM     "sentry_rulefirehistory"
WHERE    (
                  "sentry_rulefirehistory"."date_added" >= 2022-09-21 00:59:26.899530+00:00
         AND      "sentry_rulefirehistory"."date_added" < 2022-09-28 00:59:26.899530+00:00
         AND      "sentry_rulefirehistory"."rule_id" = 2)
GROUP BY "sentry_rulefirehistory"."group_id",
         (
                  SELECT   u0."event_id"
                  FROM     "sentry_rulefirehistory" u0
                  WHERE    (
                                    u0."date_added" >= 2022-09-21 00:59:26.899530+00:00
                           AND      u0."date_added" < 2022-09-28 00:59:26.899530+00:00
                           AND      u0."rule_id" = 2
                           AND      u0."group_id" = ("sentry_rulefirehistory"."group_id"))
                  ORDER BY u0."date_added" DESC limit 1)
```

Reply:
I'm not sure why it's adding the subquery to the group by, might be worth seeing if you can prevent that since it makes the query a bit less efficient. I tested this out in production and it doesn't seem too bad either way though

Reply:
could be the `OuterRef`?

Reply:
actually that seems to be basic `SubQuery` behavior
92612902712346543490144176

Reply:
using a hack I found (here)[https://code.djangoproject.com/ticket/30158] seems to remove that group by item
```
class NoGroupBySubquery(Subquery):
    def get_group_by_cols(self):
        return []
```

Reply:
Nice, it might make sense to use that here. It looks like it has been fixed in a newer Django as well. Could be worth adding a comment that this can be removed when we upgrade to Django 3.0, since the fix never made it to 2.2 https://github.com/django/django/commit/fb3f034f1c63160c0ff13c609acd01c18be12f80